### DISCUSSION TITLE
Numerically simulating CV in ODE models?

### DISCUSSION THREAD

**Discussion by MarcBerliner** [pybamm-team/PyBaMM#2025] [General]
Hi,

I was trying to figure out how PyBaMM numerically implements CV for systems of ODEs (like the SPM) but I couldn't quite figure it out. I see there's `"two-step"` and `"ode"` options for `cccv_handling`, but can you please explain what these are doing under the hood?

Thanks!

**Comment by valentinsulzer** [comment]
"two-step" adds an algebraic equation (`0 = V_sim - V_applied`) for the current density (`i`), where `V_sim` is implicitly a function of `i` so the current is adjusted to whatever it needs to be for the voltage to be satisfied. In this case there is a CC step, then a CV step. https://github.com/pybamm-team/PyBaMM/blob/5643116d5b8bd2622c13879d09b07d9952714c5b/pybamm/models/submodels/external_circuit/function_control_external_circuit.py#L81
This means the number of states in the model changes between CC and CV, and the way we actually implement the switch between CC and CV is really hacky, but in Julia you can probably do it quite neatly with an integrator-modifying callback

"ode" adds an ODE for the current that ensures you get CCCV behavior (though I haven't tested the full range of conditions in which this works). In this case there is one single CCCV step. https://github.com/pybamm-team/PyBaMM/blob/5643116d5b8bd2622c13879d09b07d9952714c5b/pybamm/models/submodels/external_circuit/function_control_external_circuit.py#L138

**Comment by MarcBerliner** [comment]
Thanks for the quick response!
